src/checkers/inference/solver/GeneralSolver.java

```diff
@@ -0,0 +1,392 @@
+
+        backEndType = backEndName == null ? BackEndType.MAXSAT : BackEndType.getBackEndType(backEndName);
+        if (backEndType == null) {
+            ErrorReporter.errorAbort("Back end \"" + backEndName + "\" has not been implemented yet.");
+        }
+
+        if (useGraph == null || useGraph.equals(Constants.TRUE)) {
+            // Configure use of constraint graph. Default is true.
+            this.useGraph = true;
+        } else {
+            this.useGraph = false;
+        }
+
+        if (backEndType.equals(BackEndType.LOGIQL)) {
+            // Configure solving strategy.
+            this.solveInParallel = false;
+        } else if (solveInParallel == null || solveInParallel.equals(Constants.TRUE)) {
+            this.solveInParallel = true;
+        } else {
+            this.solveInParallel = false;
+        }
+
+        if (collectStatistic == null || collectStatistic.equals(Constants.FALSE)) {
+            // Configure statistic collection.
+            this.collectStatistic = false;
+        } else if (collectStatistic.equals(Constants.TRUE)) {
+            this.collectStatistic = true;
+        }
+
+        // Sanitize the configuration if it needs.
+        sanitizeConfiguration();
+        System.out.println("Configuration: \nback end type: " + this.backEndType.simpleName + "; \nuseGraph: "
+                + this.useGraph + "; \nsolveInParallel: " + this.solveInParallel + ".");
+    }
+
+    protected void configureLattice(QualifierHierarchy qualHierarchy) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        lattice = latticeBuilder.buildLattice(qualHierarchy);
+    }
+
+    protected TwoQualifiersLattice createTwoQualifierLattice(AnnotationMirror top, AnnotationMirror bottom) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        return latticeBuilder.buildTwoTypeLattice(top, bottom);
+    }
+
+    /**
+     * This method create the default serializer for a given backEndType.
+     *
+     * If customized serialization logic is needed, one can override this method and
+     * return a customized serializer corresponding to the given backEndType.
+     *
+     * @param backEndType the backEndType that serializer will associate with.
+     * @param lattice the target type qualifier lattice.
+     * @return A Serializer compatible with the given backEndType.
+     */
+    protected Serializer<?, ?> createSerializer(BackEndType backEndType, Lattice lattice) {
+            return backEndType.createDefaultSerializer(lattice);
+    }
+
+    protected ConstraintGraph generateGraph(Collection<Slot> slots, Collection<Constraint> constraints,
+            ProcessingEnvironment processingEnvironment) {
+        GraphBuilder graphBuilder = new GraphBuilder(slots, constraints);
+        ConstraintGraph constraintGraph = graphBuilder.buildGraph();
+        return constraintGraph;
+    }
+
+    protected BackEnd<?, ?> createBackEnd(BackEndType backEndType, Map<String, String> configuration,
```

Comment:
Refactor `BackEndType` class, instead of using hard code enum value, think about using sub classes.